cassandra-analytics-cdc-sidecar/src/main/java/org/apache/cassandra/cdc/sidecar/SidecarStatePersister.java

```diff
@@ -0,0 +1,434 @@
+    {
+        this.sidecarCdcOptions = sidecarCdcOptions;
+        this.cdcOptions = cdcOptions;
+        this.sidecarCdcStats = sidecarCdcStats;
+        this.cassandraClient = cassandraClient;
+        this.asyncExecutor = asyncExecutor;
+    }
+
+    // StatePersister implemented methods
+
+    @Override
+    public void persist(String jobId, int partitionId, @Nullable TokenRange tokenRange, @NotNull ByteBuffer buf)
+    {
+        final PersistWrapper latest = new PersistWrapper(jobId, partitionId, tokenRange, buf, timestampGenerator.next());
+        PersistWrapper.Key key = latest.key();
+        if (!latest.equals(this.latestState.get(key)))
+        {
+            this.latestState.compute(key, (k, prev) -> !latest.equals(prev) ? latest : prev);
+        }
+    }
+
+    @NotNull
+    @Override
+    public List<CdcState> loadState(String jobId, int partitionId, @Nullable TokenRange tokenRange)
+    {
+        CompressionUtil compressionUtil = CdcBridgeFactory.get(cdcOptions.version()).compressionUtil();
+        List<Integer> sizes = new ArrayList<>();
+        // deserialize and merge the CDC state objects into canonical view
+        List<CdcState> result = loadStateForRange(jobId, tokenRange)
+                                .peek(bytes -> sizes.add(bytes.length))
+                                .map(bytes -> CdcState.deserialize(CdcKryoRegister.kryo(), compressionUtil, bytes))
+                                .collect(Collectors.toList());
+        int count = sizes.size();
+        int len = sizes.stream().mapToInt(i -> i).sum();
+        LOGGER.debug("Read CDC state from Cassandra jobId={} start={} end={} stateCount={} stateSize={}",
+                     jobId, tokenRange == null ? "null" : tokenRange.lowerEndpoint(), tokenRange == null ? "null" : tokenRange.upperEndpoint(), count, len);
+        sidecarCdcStats.captureCdcConsumerReadFromState(count, len);
+        return result;
+    }
+
+    @VisibleForTesting
+    public Stream<byte[]> loadStateForRange(String jobId, @Nullable TokenRange tokenRange)
+    {
+        return cassandraClient
+               .loadStateForRange(jobId, tokenRange);
+    }
+
+    /**
+     * Start the SidecarStatePersister to flush to Cassandra every `persistDelay()`
+     */
+    public synchronized void start()
+    {
+        if (timerId >= 0)
+        {
+            // already running
+            return;
+        }
+        this.timerId = asyncExecutor.periodicTimer(this::persistToCassandra, sidecarCdcOptions.persistDelay().toMillis());
+    }
+
+    /**
+     * Stop the SidecarStatePersister gracefully, blocking to await for any pending flushes to complete.
+     */
+    public void stop()
+    {
+        stop(true);
+    }
+
+    public synchronized void stop(boolean flush)
+    {
+        if (this.timerId < 0)
+        {
+            // not running
+            return;
+        }
+
+        asyncExecutor.cancelTimer(this.timerId);
+        this.timerId = -1;
```

Comment:
Relying on the condition `timerId == -1` might not be sufficient. The scheduled run could be still running when cancelTimer returns. However, it is not something to be addressed in this patch. Just calling it out.